Mathlib/Data/Matrix/ColumnRowPartitioned.lean

```diff
@@ -0,0 +1,229 @@
+  unfold fromColumns toColumns₁ toColumns₂
+  funext i j
+  cases' j
+  all_goals (simp only [of_apply, Sum.elim_inl, Sum.elim_inr])
+
+lemma fromColumns_ext_iff (A₁ : Matrix M N₁ R) (A₂ : Matrix M N₂ R) (B₁ : Matrix M N₁ R)
+    (B₂ : Matrix M N₂ R) : fromColumns A₁ A₂ = fromColumns B₁ B₂ ↔ A₁ = B₁ ∧ A₂ = B₂ := by
+  simp_rw [fromColumns, ← Matrix.ext_iff, of_apply, Sum.forall, Sum.elim_inl, Sum.elim_inr]
+  exact ⟨(fun h => ⟨fun i j => (h i).1 j, fun _ _ => (h _).2 _⟩),
+    (fun h => (fun _ => ⟨(h.1 _), (h.2 _)⟩))⟩
+
+lemma fromRows_toRows (A : Matrix (M₁ ⊕ M₂) N R) : A = fromRows A.toRows₁ A.toRows₂ := by
+  unfold fromRows toRows₁ toRows₂
+  funext i j
+  cases' i
+  all_goals (simp only [of_apply, Sum.elim_inl, Sum.elim_inr])
+
+lemma fromRows_ext_iff (A₁ : Matrix M₁ N R) (A₂ : Matrix M₂ N R) (B₁ : Matrix M₁ N R)
+    (B₂ : Matrix M₂ N R) : fromRows A₁ A₂ = fromRows B₁ B₂ ↔ A₁ = B₁ ∧ A₂ = B₂ := by
+  simp_rw [fromRows, ← Matrix.ext_iff, of_apply, Sum.forall, Sum.elim_inl, Sum.elim_inr]
+
+/- A column partioned matrix when transposed gives a row partioned matrix with columns of the
+initial matrix tranposed to become rows. -/
+lemma transpose_fromColumns_eq_fromRows_transpose (A₁ : Matrix M N₁ R) (A₂ : Matrix M N₂ R) :
+    transpose (fromColumns A₁ A₂) = fromRows (transpose A₁) (transpose A₂) := by
+  rw [fromColumns, fromRows]
+  funext i j
+  cases' i with i i
+  all_goals (simp only [transpose_apply, of_apply, Sum.elim_inl, Sum.elim_inr] )
+
+/- A row partioned matrix when transposed gives a column partioned matrix with rows of the initial
+matrix tranposed to become columns. -/
+lemma transpose_fromRows_eq_fromColumns_transpose (A₁ : Matrix M₁ N R) (A₂ : Matrix M₂ N R) :
+    transpose (fromRows A₁ A₂) = fromColumns (transpose A₁) (transpose A₂) := by
+  rw [fromColumns, fromRows]
+  funext i j
+  cases' j with j j
+  all_goals (simp only [transpose_apply, of_apply, Sum.elim_inl, Sum.elim_inr] )
+
+section Semiring
+
+variable [Semiring R]
+
+lemma fromRows_mul (A₁ : Matrix M₁ N R) (A₂ : Matrix M₂ N R) (B : Matrix N M R) :
+    (fromRows A₁ A₂) ⬝ B = fromRows (A₁⬝B) (A₂⬝B) := by
+  unfold fromRows
+  funext i j
+  cases' i with i i
+  all_goals (simp only [mul_apply, of_apply, Sum.elim_inl, Sum.elim_inr] )
+
+lemma mul_fromColumns (A : Matrix M N R) (B₁ : Matrix N N₁ R) (B₂ : Matrix N N₂ R) :
+    A ⬝ (fromColumns B₁ B₂) = fromColumns (A⬝B₁) (A⬝B₂) := by
+  unfold fromColumns
+  funext i j
+  cases' j with j j
+  all_goals (simp only [mul_apply, of_apply, Sum.elim_inl, Sum.elim_inr] )
+
+lemma fromRows_zero : fromRows (0 : Matrix M₁ N R) (0 : Matrix M₂ N R) = 0 := by
+  funext i j
+  unfold fromRows
+  simp only [Sum.elim_zero_zero, of_apply, Pi.zero_apply, zero_apply]
+
+lemma fromColumns_zero : fromColumns (0 : Matrix M N₁ R) (0 : Matrix M N₂ R) = 0 := by
+  funext i j
+  unfold fromColumns
+  cases' j with j j
+  all_goals (simp only [of_apply, Sum.elim_inl, Sum.elim_inr, zero_apply])
+
```

Comment:
Could you add `simp` lemmas saying `fromColumns (fromRows A₁ A₂) (fromRows B₁ B₂) = fromBlocks A₁ A₂ B₁ B₂` and `fromRows (fromColumns A₁ A₂) (fromColumns B₁ B₂) = fromBlocks A₁ A₂ B₁ B₂`?